jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PoolingPathResource.java

```diff
@@ -0,0 +1,142 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util.resource;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystemAlreadyExistsException;
+import java.nio.file.FileSystemNotFoundException;
+import java.nio.file.FileSystems;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.eclipse.jetty.util.IO;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Java NIO Path Resource with file system pooling. All {@link FileSystem} implementations must use this class
+ * except for the default one (returned by {@link FileSystems#getDefault()}).
+ */
+public class PoolingPathResource extends PathResource
+{
+    private static final Logger LOG = LoggerFactory.getLogger(PoolingPathResource.class);
+
+    private static final Map<String, ?> EMPTY_ENV = new HashMap<>();
+    private static final Map<FileSystem, AtomicInteger> POOL = new HashMap<>();
+    private static final AutoLock POOL_LOCK = new AutoLock();
+
+    private boolean closed;
+
+    PoolingPathResource(URI uri) throws IOException
+    {
+        super(createFsIfNeeded(uri), true);
+    }
+
+    private static URI createFsIfNeeded(URI uri) throws IOException
+    {
+        if (!uri.isAbsolute())
+            throw new IllegalArgumentException("not an absolute uri: " + uri);
+        if (PathResource.ALLOWED_SCHEMES.contains(uri.getScheme().toLowerCase(Locale.ROOT)))
+            throw new IllegalArgumentException("not an allowed scheme: " + uri);
```

Comment:
Is this correct?   Don't was want `jar:` and `jarfile:` resources to come via this path?

Reply:
`PathResource.ALLOWED_SCHEMES` contains the list of URI schemes that do not need to and must not be pooled. This check makes sure we do not pool `file:` or `jrt:` URIs as that's not needed and trying to close those filesystems actually is sanctioned with an exception.